### ISSUE TITLE
use higher (> 255) hwdb key codes in our multimedia keyboard tables now that X11 is dead?

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#34323]
### systemd version the issue has been seen with

git main

### CPU architectures issue was seen on

None

### Expected behaviour you didn't see

We ship a myriad of hwdb entries for multimedia keyboards, and have been for a long time. Many of these mappings avoid keycodes above 255, because X11 can only do 8bit keycodes iirc. (Right @whot?)

But given that X11 is kinda dead, and only a compat thing at this point (Right?), I wonder if we should fix this now?

Specifically, there are a bunch of entries which map "micmute" keys to "f20" rather than "micmute". Has the time come to fix this?

Of course, this would mean for X11 if they still care for this compat, that the wayland-to-x11 translation layer would start to systematically map micmute back to f20, if desired? i don't know how this works.

other relevant keys:

"f21" → "touchpad-toggle"
"f22" → "touchpad-on"

And that's it, I think.

@martinpitt you were involved in this back in the day? maybe you have something to add to this? anything i forgot about this? did i musinderstand the situation?

(as it turns out our current hwdb files already mix/match micmute and f20 all over the place. newer netries use the former, older entries tend to use the later, and its that lack of systematic definitions that makes me irk)

**Comment by poettering** [comment]
(please CC anyone else from the wayland/x11 camps, who might care about this)

**Comment by poettering** [comment]
I prepped #34325 now that addresses this. But before we do, would love some agreement that this is the right thing to do.

**Comment by whot** [comment]
> because X11 can only do 8bit keycodes iirc.

Correct, there is no reasonable way around this without rewriting large parts of XKB and getting every X client to update those new bits.


> But given that X11 is kinda dead, and only a compat thing at this point (Right?), I wonder if we should fix this now?

- First Distributions start shipping Wayland
- First Distributions start shipping Wayland by default
- First Developers allow for X to be disabled        **← you are here**
- First Distributions stop shipping X by default
- First Developers remove support for X
- First Distributions remove support for X

It's going to be a very long time until we go from "first" to "most". Right now if you ship this you'll break those keys on anything that updates systemd but doesn't run Wayland exclusively.

IMO it's a bit to early though I wonder how micmute works on those setups...

**Comment by poettering** [comment]
why can't the wayland-to-x11 translation layer map kernel KEY_MICMUTE to KEY_F20 before passing it on? i.e. it sounds weird to me to lie to wayland if the only thing that needs lying to is x11, and it could as well be wayland that lies to x11 clients when it passes on the events?

**Comment by whot** [comment]
Because for the rather many users of Xorg there is no wayland-to-x11 translation layer :) 

The only way to map this would be adding those hacks into the xf86-input-evdev/libinput drivers. I guess I'll PRs for that, with all the ugliness that implies...

**Comment by poettering** [comment]
> Because for the rather many users of Xorg there is no wayland-to-x11 translation layer :)

ah, i thought x11 directly on drivers was not supported anymore.

> The only way to map this would be adding those hacks into the xf86-input-evdev/libinput drivers. I guess I'll PRs for that, with all the ugliness that implies...

maybe add a flag for libinput context creation DONT_MAP_KEYCODES_TO_8BIT or so. Then, set that flag when wayland consumes the lib, but leave it unset when x11 does?

**Comment by poettering** [comment]
i mean, some hwdb entries started to use micmute instead of f20 recently. 

i think it would be nicer to move to clean hwdb files, and leave any compat mapping to further up the stack. 

but ultimately i would defer to you on this. please decide what the way forward shall be for systemd:

1. start using micmute/touchpad-toggle/touchpad-on/touchpad-off literally, as implemented by https://github.com/systemd/systemd/pull/34325.
2. fix current uses of these keys in hwdb, and patch them to also return f20,f21,… and delay things for a couple of more years.

I personally would prefer solution 1 of course. But if you'd prefer 2 we can do that too. Just let me know.

**Comment by whot** [comment]
> i think it would be nicer to move to clean hwdb files, and leave any compat mapping to further up the stack.

I agree. Just give me a bit of time to fix this in the xorg drivers please. Ping me in... 4 weeks? (Holiday)

**Comment by poettering** [comment]
> I agree. Just give me a bit of time to fix this in the xorg drivers please. Ping me in... 4 weeks? (Holiday)

okeydokey! thanks!

enjoy your holiday!

**Comment by jwrdegoede** [comment]
@whot for adding the compatibility mappings to xorg-x11-drv-libinput I'm aware of the following mappings being necessary:

- KEY_MICMUTE -> KEY_F20
- KEY_TOUCHPAD_TOGGLE -> KEY_F21
- KEY_TOUCHPAD_ON -> KEY_F22
- KEY_TOUCHPAD_OFF-> KEY_F23

These are from: /usr/share/X11/xkb/symbols/inet:

```
    key <FK21>   {      [ XF86TouchpadToggle    ]       };
    key <FK22>   {      [ XF86TouchpadOn        ]       };
    key <FK23>   {      [ XF86TouchpadOff       ]       };

    key <FK20>   {      [ XF86AudioMicMute      ]       };
```


There are also 3 mappings to F24 in hwdb:

```
evdev:atkbd:dmi:bvn*:bvr*:bd*:svnAcer*:pnTravelMate*C3[01]0*:*
 KEYBOARD_KEY_67=f24                                    # FIXME: rotate screen

evdev:name:ThinkPad Extra Buttons:dmi:bvn*:bvr*:bd*:svnIBM*:pn*:*
 KEYBOARD_KEY_08=f24                                    # Fn+F9 undock

evdev:name:ThinkPad Extra Buttons:dmi:bvn*:bvr*:bd*:svnLENOVO*:pn*:*
 KEYBOARD_KEY_08=f24
```

These are weird, looking at xkb, keycodes/evdev maps KEY_F24 -> FK24, but then grepping for FK24 in symbols shows that FK24 is only mapped in symbols/fujitsu_vndr/us suggesting that these hwdb mappings lead to there being no X11 key-symbol associated with the key ?

The FIXME: rotate-screen one is probably because of this:

```
input-event-codes.h:
#define KEY_ROTATE_DISPLAY      153     /* Display orientation for e.g. tablets */
symbols/init:
    key <I161>   {      [ XF86RotateWindows     ]       };
```

So xkb is mapping KEY_ROTATE_DISPLAY to XF86RotateWindows, otoh there are other hwdb entries which are happily using rotate_display I guess these assume that desktop-environment map XF86RotateWindows to display-rotation ?

I'm not sure what to do with these. Either change them all to map to rotate_display, or ... ?

Also maybe we should add:

```
key <FK19> { [ F19		]	};
key <FK24> { [ F24		]	};
```

to symbols/inet? because currently these are unmapped.

**Comment by poettering** [comment]
> These are weird, looking at xkb, keycodes/evdev maps KEY_F24 -> FK24, but then grepping for FK24 in symbols shows that FK24 is only mapped in symbols/fujitsu_vndr/us suggesting that these hwdb mappings lead to there being no X11 key-symbol associated with the key ?

i have the suspicion people mapped these three to f24 simply because there was nothing better at that time. 

there first one i changed to rotate_display in the other PR btw, trusting the commentary.

the undock thing probably should be mapped to some new KEY_UNDOCK in an ideal world. But then again, this seems to be for historic laptops and not sure anyone really cares who still has that hw.

**Comment by jwrdegoede** [comment]
> > These are weird, looking at xkb, keycodes/evdev maps KEY_F24 -> FK24, but then grepping for FK24 in symbols shows that FK24 is only mapped in symbols/fujitsu_vndr/us suggesting that these hwdb mappings lead to there being no X11 key-symbol associated with the key ?
> 
> I have the suspicion people mapped these three to f24 simply because there was nothing better at that time.
> 
> there first one i changed to rotate_display in the other PR btw, trusting the commentary.
> 
> the undock thing probably should be mapped to some new KEY_UNDOCK in an ideal world. But then again, this seems to be for historic laptops and not sure anyone really cares who still has that hw.

I agree with changing the first one to rotate_display. The 2 ThinkPad ones should probably just be left at f24. Some ThinkPad users have highly tweaked / customized setups and if we change those we run the risk of running into: https://xkcd.com/1172/

Since we have nothing better to map those to just leaving them alone seems best.

**Comment by martinpitt** [comment]
Hello all! Sorry for late response, just returned from holidays.

@poettering you did understand this correctly. I also think it's a bit early, but I like the approach of mapping the keys back to Fxx in xorg-x11-drv-libinput. I suggest to block #34325 on that.

> There are also 3 mappings to F24 in hwdb:

Unlike F20 to 23, that didn't have a standard interpretation in X.org. The idea was to map these rare keys to *something* so that users can set up a custom key binding in GNOME and then do whatever they like. For `rotate_display` we should just move to that official key code, and I have absolutely no idea what an "undock" key would/should do, or if it even exists now. Every laptop that I've seen in the last 20 years undocks by pulling a lever or a mechanical button, one doesn't have  to pre-announce it via a key press. So perhaps it's best to just leave that as F24 in the spirit of space bar heating :grin: 

Thanks!

**Comment by whot** [comment]
> lso maybe we should add:
> ```
>    key <FK19> { [ F19		]	};
>    key <FK24> { [ F24		]	};
> ```
> to symbols/inet? because currently these are unmapped.

There's [`-option fkeys:fkeys`](https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/merge_requests/726) now. I'd rather not provide any new standard mappings for F-keys since all the others are already mapped to special symbols and it's a bit weird if F19/F24 do something normal but F20 and F23 shut down your mic/touchpad.

I've filed [xf86-input-libinput!62](https://gitlab.freedesktop.org/xorg/driver/xf86-input-libinput/-/merge_requests/62) now for the mapping, so IMO this PR here can be merged.

**Comment by whot** [comment]
I'll cut a release for both next week, pls ping me if I end up forgetting somehow.

**Comment by whot** [comment]
> I'll cut a release for both next week,

done - please feel free to go nuts (with the code changes, not out of excitement about a new evdev release, that'd be weird ;)
